### ISSUE TITLE
Rustdoc: Unable to fulfill trait

### ISSUE THREAD

**Issue by OussamaElgoumri** [rust-lang/rust#53102]
<short summary of the bug>

**I tried this code:**

cargo doc

**I expected to see this happen:** documentation

**Instead, this happened:** 
 Documenting capnp v0.8.17
    Checking iovec v0.1.2
    Checking net2 v0.2.33
    Checking rand v0.4.2
    Checking num_cpus v1.8.0
    Checking memchr v2.0.1
    Checking time v0.1.40
thread '<unnamed>' panicked at 'Unable to fulfill trait DefId(2/0:865 ~ core[a6fc]::marker[0]::Send[0]) for 'list_list::Owned<T>': [FulfillmentError(Obligation(predicate=Binder(ProjectionPredicate(ProjectionTy { substs: [T, ReStatic], item_def_id: DefId(0/0:1150 ~ capnp[8787]::traits[0]::Owned[0]::Reader[0]) }, _)),depth=2),Ambiguity), FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<_ as std::marker::Send>)),depth=2),Ambiguity)]', librustc/traits/auto_trait.rs:218:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.
   Compiling openssl-sys v0.9.33

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.28.0 (9634041f0 2018-07-30) running on x86_64-unknown-linux-gnu

error: Could not document `capnp`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name capnp /home/oussama/.cargo/registry/src/github.com-1ecc6299db9ec823/capnp-0.8.17/src/lib.rs -o /home/oussama/Desktop/rust/playground/target/doc -L dependency=/home/oussama/Desktop/rust/playground/target/debug/deps --extern byteorder=/home/oussama/Desktop/rust/playground/target/debug/deps/libbyteorder-af85c17f60f27595.rmeta` (exit code: 101)
warning: build failed, waiting for other jobs to finish...
error: build failed

## Meta

`rustc --version --verbose`:

rustc 1.28.0 (9634041f0 2018-07-30)
binary: rustc
commit-hash: 9634041f0e8c0f3191d2867311276f19d0a42564
commit-date: 2018-07-30
host: x86_64-unknown-linux-gnu
release: 1.28.0
LLVM version: 6.0

**Comment by dwrensha** [comment]
`cargo doc -p capnp` works without any trouble on rustc 1.25.0, and hits the above error on rustc 1.26.0, so the problem must have been introduce in that release cycle.

**Comment by dwrensha** [comment]
Smaller piece of code that hits the same problem:
```rust
// lib.rs

pub trait FromPointerReader<'a> : Sized {}

pub struct PointerReader<'a>  {
    _arena: &'a [u8],
}

pub trait OwnedTrait<'a> {
    type Reader: FromPointerReader<'a>;
}

pub struct Owned<T> where T: for<'a> OwnedTrait<'a> {
    marker: ::std::marker::PhantomData<<T as OwnedTrait<'static>>::Reader>,
}
```

```
$ rustdoc --version
rustdoc 1.28.0 (9634041f0 2018-07-30)
$ rustdoc lib.rs
thread '<unnamed>' panicked at 'Unable to fulfill trait DefId(2/0:865 ~ core[537d]::marker[0]::Send[0]) for 'Owned<T>': [FulfillmentError(Obligation(predicate=Binder(ProjectionPredicate(ProjectionTy { substs: [T, ReStatic], item_def_id: DefId(0/0:6 ~ lib[8787]::OwnedTrait[0]::Reader[0]) }, _)),depth=2),Ambiguity), FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<_ as std::marker::Send>)),depth=2),Ambiguity)]', librustc/traits/auto_trait.rs:218:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.28.0 (9634041f0 2018-07-30) running on x86_64-apple-darwin
```

**Comment by dtolnay** [comment]
Fixed in #52990.
